peer/peer.go

```diff
@@ -0,0 +1,1942 @@
+	}
+	na := wire.NewNetAddressIPPort(ip, uint16(port), services)
+	return na, nil
+}
+
+// outMsg is used to house a message to be sent along with a channel to signal
+// when the message has been sent (or won't be sent due to things such as
+// shutdown)
+type outMsg struct {
+	msg      wire.Message
+	doneChan chan struct{}
+}
+
+// Stats is the collection of stats related to a peer
+type Stats struct {
+	StatsMtx           sync.Mutex // protects all statistics below here.
+	versionKnown       bool
+	protocolVersion    uint32
+	versionSent        bool
+	verAckReceived     bool
+	services           wire.ServiceFlag
+	timeOffset         int64
+	timeConnected      time.Time
+	lastSend           time.Time
+	lastRecv           time.Time
+	bytesReceived      uint64
+	bytesSent          uint64
+	userAgent          string
+	startingHeight     int32
+	lastBlock          int32
+	lastAnnouncedBlock *wire.ShaHash
+	lastPingNonce      uint64    // Set to nonce if we have a pending ping.
+	lastPingTime       time.Time // Time we sent last ping.
+	lastPingMicros     int64     // Time for last ping to return.
+}
+
+// ShaFunc is a function which returns a block sha, height and error
+// It is used as a callback to get newest block details.
+type ShaFunc func() (sha *wire.ShaHash, height int64, err error)
+
+// Peer provides a bitcoin peer for handling bitcoin communications.  The
+// overall data flow is split into 3 goroutines and a separate block manager.
+// Inbound messages are read via the inHandler goroutine and generally
+// dispatched to their own handler.  For inbound data-related messages such as
+// blocks, transactions, and inventory, the data is passed on to the block
+// manager to handle it.  Outbound messages are queued via QueueMessage or
+// QueueInventory.  QueueMessage is intended for all messages, including
+// responses to data such as blocks and transactions.  QueueInventory, on the
+// other hand, is only intended for relaying inventory as it employs a trickling
+// mechanism to batch the inventory together.  The data flow for outbound
+// messages uses two goroutines, queueHandler and outHandler.  The first,
+// queueHandler, is used as a way for external entities (mainly block manager)
+// to queue messages quickly regardless of whether the peer is currently
+// sending or not.  It acts as the traffic cop between the external world and
+// the actual goroutine which writes to the network socket.  In addition, the
+// peer contains several functions which are of the form pushX, that are used
+// to push messages to the peer.  Internally they use QueueMessage.
+type Peer struct {
+	btcnet             wire.BitcoinNet
+	started            int32
+	connected          int32
+	disconnect         int32 // only to be used atomically
+	conn               net.Conn
+	addr               string
+	na                 *wire.NetAddress
+	id                 int32
+	inbound            bool
+	persistent         bool
+	KnownAddresses     map[string]struct{}
```

Comment:
I think we'll need an update function that is protected by a mutex  for that.  Raw unprotected access to a map isn't concurrent safe.